packages/edit-site/src/components/editor/global-styles-provider.js

```diff
@@ -58,12 +57,75 @@ export const useGlobalStylesReset = () => {
 	];
 };
 
+const extractSupportKeys = ( supports, metadata ) => {
+	const supportKeys = [];
+	Object.keys( metadata.properties ).forEach( ( name ) => {
+		if ( get( supports, metadata.properties[ name ].support, false ) ) {
+			supportKeys.push( name );
+		}
+	} );
+	return supportKeys;
+};
+
+const getContexts = ( blockTypes, metadata ) => {
+	const contexts = {};
+
+	// Add contexts from block metadata.
+	blockTypes.forEach( ( blockType ) => {
+		const blockName = blockType.name;
+		const blockSelector = blockType?.supports?.__experimentalSelector;
+		const supports = extractSupportKeys( blockType?.supports, metadata );
+		const hasSupport = supports.length > 0;
+
+		if ( hasSupport && typeof blockSelector === 'string' ) {
+			contexts[ blockName ] = {
+				selector: blockSelector,
+				supports,
+				blockName,
+			};
+		} else if ( hasSupport && typeof blockSelector === 'object' ) {
+			Object.keys( blockSelector ).forEach( ( key ) => {
+				contexts[ key ] = {
+					selector: blockSelector[ key ].selector,
+					supports,
+					blockName,
+					title: blockSelector[ key ].title,
+					attributes: blockSelector[ key ].attributes,
+				};
+			} );
+		} else if ( hasSupport ) {
+			const suffix = blockName.replace( 'core/', '' ).replace( '/', '-' );
+			contexts[ blockName ] = {
+				selector: '.wp-block-' + suffix,
+				supports,
+				blockName,
+			};
+		}
+	} );
+
+	// Add contexts provided by the metadata.
+	Object.keys( metadata.contexts ).forEach( ( key ) => {
+		contexts[ key ] = metadata.contexts[ key ];
+	} );
+
+	return contexts;
+};
+
 export default function GlobalStylesProvider( {
 	children,
 	baseStyles,
-	contexts,
+	metadata,
 } ) {
 	const [ content, setContent ] = useGlobalStylesEntityContent();
+	const { blockTypes, settings } = useSelect( ( select ) => {
+		return {
+			blockTypes: select( 'core/blocks' ).getBlockTypes(),
+			settings: select( 'core/edit-site' ).getSettings(),
+		};
+	} );
+	const { updateSettings } = useDispatch( 'core/edit-site' );
+
+	const contexts = getContexts( blockTypes, metadata );
```

Comment:
We are generating a new contexts reference each time the component rerenders. I guess we could just add a useMemo in this case.

Reply:
fixed at https://github.com/WordPress/gutenberg/pull/27099/commits/d35abc731519712087ee88ea75a74102d5955ee4